arrows/klv/klv_read_write.cxx

```diff
@@ -0,0 +1,170 @@
+// This file is part of KWIVER, and is distributed under the
+// OSI-approved BSD 3-Clause License. See top-level LICENSE file or
+// https://github.com/Kitware/kwiver/blob/master/LICENSE for details.
+
+/// \file
+/// Implementation of the non-templated basic KLV read/write functions.
+
+#include "klv_read_write.txx"
+
+namespace kv = kwiver::vital;
+
+namespace kwiver {
+
+namespace arrows {
+
+namespace klv {
+
+// ---------------------------------------------------------------------------
+uint64_t
+_imap_pos_infinity( size_t length )
+{
+  return length ? 0xC8ull << ( ( length - 1 ) * 8 ) : 0;
+}
+
+// ---------------------------------------------------------------------------
+uint64_t
+_imap_neg_infinity( size_t length )
+{
+  return length ? 0xE8ull << ( ( length - 1 ) * 8 ) : 0;
+}
+
+// ---------------------------------------------------------------------------
+uint64_t
+_imap_pos_quiet_nan( size_t length )
+{
+  return length ? 0xD0ull << ( ( length - 1 ) * 8 ) : 0;
+}
+
+// ---------------------------------------------------------------------------
+uint64_t
+_imap_neg_quiet_nan( size_t length )
+{
+  return length ? 0xF0ull << ( ( length - 1 ) * 8 ) : 0;
+}
+
+// ---------------------------------------------------------------------------
+uint64_t
+_imap_pos_signal_nan( size_t length )
+{
+  return length ? 0xD8ull << ( ( length - 1 ) * 8 ) : 0;
+}
+
+// ---------------------------------------------------------------------------
+uint64_t
+_imap_neg_signal_nan( size_t length )
+{
+  return length ? 0xF8ull << ( ( length - 1 ) * 8 ) : 0;
+}
+
+// ---------------------------------------------------------------------------
+_imap_terms
+_calculate_imap_terms( double minimum, double maximum, size_t length )
+{
+  // ST1201, Section 8.1.2
+  auto const float_exponent = std::ceil( std::log2( maximum - minimum ) );
+  auto const int_exponent = 8.0 * length - 1.0;
+
+  _imap_terms result = {};
+  result.forward_scale = std::exp2( int_exponent - float_exponent );
+  result.backward_scale = std::exp2( float_exponent - int_exponent );
+  result.zero_offset = ( minimum < 0 && maximum > 0 )
+                       ? result.forward_scale * minimum -
+                       std::floor( result.forward_scale * minimum )
+                       : 0.0;
+  return result;
+}
+
+// ---------------------------------------------------------------------------
+size_t
+klv_imap_length( double minimum, double maximum, double precision )
+{
+  // ST1201, Section 8.1.1
+  _check_range_precision( minimum, maximum, precision );
+
+  auto const length_bits = std::ceil( std::log2( maximum - minimum ) ) -
+                           std::floor( std::log2( precision ) ) + 1;
```

Comment:
Weak suggestion, but possibly more clear that we're still dealing with floating point numbers in this expression...
```suggestion
                           std::floor( std::log2( precision ) ) + 1.0;
```